[docs] Accessibility in Base UI (#39264)
mui/material-ui: docs/data/base/getting-started/accessibility/accessibility.md

# Accessibility

<p class="description">
Accessibility is our top priority for Base UI components.
</p>

The library, however cannot make your application fully accessible on its own.
This page highlights the most common actions you, as an application developer, need to take to ensure your product is available to use by people with disabilites.

## Keyboard navigation

As the Base UI components follow the [WAI-ARIA 1.2 standard](https://www.w3.org/TR/wai-aria-1.2/), they are accessible with a keyboard.
This is important for users who have trouble using a pointing device, but also comes in handy for users who find navigating with a keyboard faster and expect web components to behave the same way as native operating system controls.

Specifically, interactive components can be focused using the <kbd>Tab</kbd> key.
List-like components (such as Select, Menu, etc.) can be browsed and activated using arrow keys, <kbd>Home</kbd>, <kbd>End</kbd>, <kbd>Enter</kbd>/<kbd>Return</kbd> and <kbd>Escape</kbd>.
The Select and Menu also let users navigate to options that begin with a provided string by pressing alphanumeric keys.

## Focus ring

> While Base UI components handle keyboard navigation, it's the developers' responsibility to indicate when a component is focused and can receive keyboard input.

Reviewer: singular works just as well and is simpler here

```suggestion
While Base UI components handle keyboard navigation, it's the developer's responsibility to indicate when a component is focused and can receive keyboard input.
```

This is usually done by styling the [`:focus`](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus) or [`:focus-visible`](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible) pseudoclasses.

> The [WCAG guidelines](https://www.w3.org/WAI/WCAG22/Understanding/focus-appearance-minimum) specify the required appearance of focused elements.

Reviewer: To make this sentence more actionable and the link text more informative

```suggestion
Read the [WCAG guidelines on focus appearance](https://www.w3.org/WAI/WCAG22/Understanding/focus-appearance-minimum) to learn more about requirements for focus indicators.
```

## Color contrast

> As Base UI controls are unstyled, it's up to you to ensure appropriate contrast between text and background.

Reviewer: ```suggestion
Since Base UI components are unstyled, it's up to you to implement the appropriate amount of color contrast between text and background.
```

> You can use online tools such as https://cliambrown.com/contrast/ to measure contrast between colors (note that it uses the [APCA algorithm](https://ruitina.com/apca-accessible-colour-contrast/) that's being considered for the WCAG 3 guidelines).

Reviewer: ```suggestion
You can use [C. Liam Brown's Color Contrast Tool](https://cliambrown.com/contrast/) to measure contrast between colors in your design and ensure that they meet the minimum requirements. 
(Note that this tool uses the [APCA algorithm](https://ruitina.com/apca-accessible-colour-contrast/), which is slated to become the new standard in WCAG 3.)
```

## Accessible labels

> It is also your responsibility to ensure that components have accessible names.

Reviewer: I would drop "also" - can't assume the reader has actually read prior sections

```suggestion
It's your responsibility to ensure that components have accessible names.
```

> For form controls (such as the [Input](/base-ui/react-input/), [Number Input](/base-ui/react-number-input/), [Select](/base-ui/react-select/), etc.), this requires adding an associated `<label>` element or placing an `aria-label` attribute on the component itself:

Reviewer: ```suggestion
For form controls (such as the [Input](/base-ui/react-input/), [Number Input](/base-ui/react-number-input/), and [Select](/base-ui/react-select/)), this requires adding an associated `<label>` element, or placing an `aria-label` attribute on the component itself:
```

# Accessibility

<p class="description">
Learn how to make the most of Base UI's accessibility features and guidelines.
</p>

Accessibility is our top priority for Base UI components.
However, the library cannot make your application fully accessible on its own.
This page highlights the most common actions that you as a developer must take to ensure that your app is accessible to users who require assistive technologies.

## Keyboard navigation

Base UI components follow the [WAI-ARIA 1.2 standard](https://www.w3.org/TR/wai-aria-1.2/), so they are accessible with a keyboard out of the box.
This is important for users who have trouble using a pointing device, but also comes in handy for users who find navigating with a keyboard faster, and who expect web components to behave the same way as native operating system controls.

Specifically, interactive components can be focused using the <kbd>Tab</kbd> key.

> List-like components (such as Select and Menu) can be browsed and activated using arrow keys, <kbd>Home</kbd>, <kbd>End</kbd>, <kbd>Enter</kbd>/<kbd>Return</kbd>, and <kbd>Escape</kbd>.

Reviewer 2: ```suggestion
List-like components (such as Select and Menu) can be browsed and activated using arrow keys, <kbd class="key">Home</kbd>, <kbd class="key">End</kbd>, <kbd class="key">Enter</kbd>/<kbd class="key">Return</kbd>, and <kbd class="key">Escape</kbd>.
```

Adding the `class="key"` declaration as I noticed the styles were off on the deploy preview. 😬 
Sorry for not including this in the previous review!

The Select and Menu also let users navigate to options that begin with a provided string by pressing alphanumeric keys.

> 

Reviewer 2: Maybe we should add a demo with both the Select and Menu here to demonstrate what we described?

## Focus ring

While Base UI components handle keyboard navigation, it's the developer's responsibility to indicate when a component is focused and can receive keyboard input.
This is usually done by styling the [`:focus`](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus) or [`:focus-visible`](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible) pseudoclasses.
Read the [WCAG guidelines on focus appearance](https://www.w3.org/WAI/WCAG22/Understanding/focus-appearance-minimum) to learn more about requirements for focus indicators.

> 

Reviewer 2: Here, too — a simple component like the Switch would be enough to demo the focus-visible treatment?

Author: Switch is the worst component to show is this case as its focus-visible is a bit messed up (the `.Mui-focusVisible` is set on a different element that the actual `:focus-visible` state). I'll use a Slider instead.

## Color contrast

Since Base UI components are unstyled, it's up to you to implement the appropriate amount of color contrast between text and background.
You can use [C. Liam Brown's Color Contrast Tool](https://cliambrown.com/contrast/) to measure contrast between colors in your design and ensure that they meet the minimum requirements.
(Note that this tool uses the [APCA algorithm](https://ruitina.com/apca-accessible-colour-contrast/), which is slated to become the new standard in WCAG 3.)

> 

Reviewer 2: And lastly, here, too! I think we could display a button with poor vs. contrast treatment.
